Core/Src/main.c

```diff
@@ -1,181 +1,223 @@
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+//  Set CS line to ON
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+//  Start the timer
+  HAL_TIM_Base_Start (&htim1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1) {
+//	  Toggle CS line
+	  HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8);
+
+	  uint8_t transmit_data[10];
+	  uint8_t receive_data[10];
+
+//	  Start bit
+	  transmit_data[0] = 1;
+//	  Single-ended, CH0
+	  transmit_data[1] = 0b10000000;
+//	  Doesn't matter
+	  transmit_data[2] = 0;
+//	  transmit and receive
+	  HAL_SPI_TransmitReceive(&hspi1, &transmit_data, &receive_data, 3, 1000);
+
+//	  Toggle CS line
+	  HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8);
+
+//	  Concatenate the last two bytes
+	  uint16_t combined = receive_data[1] << 8 | receive_data[2];
+//	  Keep the last 10 bits
+	  uint16_t adc_output = 0x03FF & combined; // 0x0000 0011 1111 1111 (only need last 10 bits)
+
+//	  Take ADC output between 0-1024 and converts to counts between 3200 and 6400 (5-10%)
+	  int counts_per_cycle = adc_output * 3.125 + 3200;
+
+//	  convert to PWM
+	  TIM_OC_InitTypeDef htim_oc_init;
+	  htim_oc_init.OCMode = TIM_OCMODE_PWM1;
+	  htim_oc_init.Pulse = (uint32_t) counts_per_cycle;
+
+	  HAL_TIM_PWM_ConfigChannel(&htim1, &htim_oc_init, TIM_CHANNEL_1);

```

Comment:
Honestly, not sure if this will work or not...

But this can all be condensed to a one line function. The timer peripheral has this register called the "Capture Compare Register" which compares the timer value to the "compare" value you provide which essentially sets the duty cycle of your PWM signal.